ethereum/contracts/TablelandRigs.sol

```diff
@@ -354,6 +390,231 @@ contract TablelandRigs is
+                tokenId, ownerOf(tokenId), block.number
+            )
+         */
+        TablelandDeployments.get().runSQL(
+            address(this),
+            _rigPilotSessionsTableId,
+            SQLHelpers.toInsert(
+                RIG_PILOT_SESSIONS_PREFIX,
+                _rigPilotSessionsTableId,
+                "rig_id,owner,start_time",
+                string.concat(
+                    StringsUpgradeable.toString(uint64(tokenId)),
+                    ",",
+                    StringsUpgradeable.toHexString(ownerOf(tokenId)),
+                    ",",
+                    StringsUpgradeable.toString(uint64(block.number))
+                )
+            )
+        );
+        emit Training(tokenId);
+    }
+
+    using ERC165Checker for address;
+
+    /**
+     * @dev See {ITablelandRigs-pilotRig}.
+     */
+    function pilotRig(
+        uint256 tokenId,
+        address pilotContract,
+        uint256 pilotTokenId
+    ) external {
+        // Check the Rig `tokenId` exists
+        if (!_exists(tokenId)) revert OwnerQueryForNonexistentToken();
+        // Verify `msg.sender` is authorized to pilot the specified Rig
+        if (ownerOf(tokenId) != _msgSenderERC721A())
+            revert InvalidRigOwnership(tokenId);
+        // Validate if pilot training is completed
+        // Once training begins, the Rig is prevented from being parked until
+        // training has been completed; thus, only need to check if the Rig has
+        // been ungaraged yet since parking handles training logic (see `parkRig` for more details)
+        if (_pilots[tokenId].index == 0) revert RigIsNotTrained(tokenId);
+        // Check if Rig is currently in-flight (not parked)
+        if (_pilots[tokenId].startTime > 0) revert RigIsNotParked(tokenId);
+        // Check if `pilotContract` is an ERC721; cannot be the Rigs contract
+        // TODO is this the right way to handle checking if the ERC721 is the Rigs contract? Or, is there a different way w/upgradeable contracts
+        if (
+            !pilotContract.supportsInterface(type(IERC721).interfaceId) ||
+            (pilotContract.supportsInterface(type(IERC721).interfaceId) &&
+                pilotContract == address(this))
+        ) revert InvalidPilotContract(pilotContract);
+        // Check ownership of `pilotTokenId` at target `pilotContract`
+        if (IERC721(pilotContract).ownerOf(pilotTokenId) != _msgSenderERC721A())
+            revert InvalidPilotOwnership(pilotTokenId);
+        // Check if the pilot is actively piloting another Rig, and if so, park the other Rig
+        for (uint256 i = 1; i <= maxSupply; i++) {
+            if (
+                // `_rigsStatus` starts at index `0`, while `tokenId` in `_pilots` starts at `1`
+                _rigsStatus[i - 1] == 1 &&
+                (_pilots[i].pilotContract == pilotContract &&
+                    _pilots[i].pilotId == pilotTokenId)
+            ) {
+                parkRig(i);
+            }
+        }
```

Comment:
is this the bulk of the gas in this method? what about using two maps:

```
mapping(uint256 => uint16) internal _pilots; // pilot -> tokenid
mapping(uint16 => Pilot) internal _rigPilots; // tokenid -> pilot data
struct Pilot {
    uint16 index;
    uint64 startTime;
    uint256 pilot; // value stored twice
}
```

or

```
mapping(uint16 => uint256) internal _rigPilots; // tokenid -> pilot
mapping(uint256 => Pilot) internal _pilots; // pilot -> pilot data
struct Pilot {
    uint16 index;
    uint64 startTime;
    uint16 rigId;
}
```

both duplicate data, but second one duplicates less. 

used use `_rigPilots` to lookup a rig's pilot, and `_pilots` to look up whether or not a pilot has been used, and to get its data (start time, index)

Reply:
maybe clear already, but for an example of packing / unpacking multiple ints into a larger int, see `getClaimed` and `_setClaimed`:

<img width="590" alt="Screen Shot 2022-10-15 at 11 18 34 AM" src="https://user-images.githubusercontent.com/361000/196002176-0856cac0-b37f-421a-940d-5df3cd202d46.png">

packs two `uint16`s into a `uint64`

Reply:
Hrm, I think it needs to be the first flavor, since we can't index pilots that are the trainer in `mapping(uint256 => Pilot) internal _pilots;`

So, expanding on the first flavor, which can be (changing names to be more clear):

```
struct Pilot {
    uint64 startTime;
    uint192 pilot; // value is indexed by _pilotIindex
}

mapping(uint16 => Pilot) internal _pilots; // tokenid -> pilot data
mapping(uint192 => uint16) internal _pilotIndex; // packed pilot -> tokenid
```

...where "pilot" is pilot contract address (`uint160`) + pilot token id (`uint32`). `uint32` allows token ids up to `4,294,967,295`, which sounds like plenty to me. You could do a check on token id to see if it'll fit in `uint32`, and error if not.

Training should be as cheap as possible to get people started in the game. With this setup, training need not touch `_pilotIndex`, just `_pilots`. During training, we could say that `pilot` is `1`.

I think we can remove `index` entirely by saying that `_pilots` stores the "latest pilot data". ie, don't delete from `_pilots` on park, just flip `startTime` to `0`. Then you can walk through some logic to determine state: 

Possible states:
1. not trained
2. in training
3. trained and parked
4. trained and piloted

- It's possible to go from (2) to (1) by "parking" while training, ie, rig must start from zero again.
- It's possible to go from (4) to (3)—that's what "parking" does when already trained.

You might consider creating a method that returns "garage status" as an enum to make logic in methods a bit cleaner. Given the possible states, the logic of this method would look something like:

```
enum GarageStatus {
    UNTRAINED,
    TRAINING,
    PARKED,
    PILOTED
}

function _pilotStatus() private returns (GarageStatus) {
    Pilot p = _pilots[uint16(tokenId)];
    if (p.startTime == 0) {
        if (p.pilot == 0) return UNTRAINED;
        // The `park` logic should set `pilot` to `0` if both of these are true:
        //   - `pilot` is currently `1`
        //   - rigs has not been training for long enough
        // ie, `park` results in a status of UNTRAINED or PARKED
        // In no other cases should `park` update `pilot` (it's either current or last used pilot)
        else return PARKED;
    } else {
        // Invariant: pilot cannot be `0` if `startTime` is > `0`.
        // ie, if rig is training or being piloted, pilot is either the trainer or a real pilot
        if (p.pilot == 1) return TRAINING;
        else return PILOTED;
    }
}

function _canPilot() private returns (bool) {
    // two ways to pilot:
    //   1. in PARKED status
    //   2. in TRAINING status for long enough
    // Invariant: you cannot switch real pilots mid-flight, but you can go
    // from trainer -> real pilot to make it easier + cheaper to pilot (continue playing the game)
    Pilot p = _pilots[uint16(tokenId)];
    if (p.startTime == 0) {
        // Rig is either UNTRAINED or PARKED. It's not possible to pilot from UNTRAINED.
        return p.pilot != 0; // could be trainer (`1`) or real pilot
    } else if (p.pilot == 1) {
        return block.number >= p.startTime + 172800;
    } else return false;
}
```

Reply:
The only function of `_pilotIndex` here is to easily check if a custom pilot is in use.